### ISSUE TITLE
[benchmarks] CPU-Bound Microbenchmark

### ISSUE THREAD

**Issue by lemosep** [ppenna/nanvix#330]
This issue aims to provide a closer description on how the CPU-bound microbenchmark will be executed. This was previously discussed on #310 . The experiment will work as follows:

- Execute a 512 x 512 matrix multiplication code.
- Analyze `vm_run` log to measure execution time.

Furthermore, I expect to be able to use `no_std` benchmark tools and hardware monitors to extract more information from the experiment.

**Comment by ppenna** [comment]
@lemosep could you clarify the following points?
- Why have you cherry-picked 512?
- Will it be an integer or float-point multiplication?
- How will you generate the input matrices?
- Which matrix multiplication algorithm will you use?

**Comment by lemosep** [comment]
- I've chosen 512x512 as a way of balancing memory usage and execution time, so it could be measurable.
- It's a `f32` multiplication
- I created an extern program using the `rand` crate.
- I've used the iterative algorithm.

**Comment by ppenna** [comment]
> I've chosen 512x512 as a way of balancing memory usage and execution time, so it could be measurable.

This sounds arbitrary. Have you based this decision on any paper? Anyhow, I suggest you make it a parameter.

> It's a f32 multiplication

Could you switch to integer multiplication instead? Floating point airhmetic also benchmarks FPU capabilities of the guest and may make it harder to interpret results.

> I created an extern program using the rand crate.

You should generate input on the program itself, based on the size of the input matrix. Say that one wants to study a different case, would you require that person to write their own matrix generator as well?

> I've used the iterative algorithm.

Sure, but are you considering using any cache-oblivious technique?

**Comment by lemosep** [comment]
> Sure, but are you considering using any cache-oblivious technique?

I'll study this approach.

**Comment by lemosep** [comment]
>You should generate input on the program itself, based on the size of the input matrix. Say that one wants to study a different case, would you require that person to write their own matrix generator as well?

Just to clarify, does the matrix generation has to be done during the execution of the benchmark? **If so**, that would mess up the benchmark measurements with additional execution time that is non-related to the multiplication.

**If not**: I'm using the `rng_core` crate to implement my own RNG, since using the `rand` crate on `no_std` environments is very limited.  The thing is, PRNG will generate the same sequence of numbers each time wit the same seed, causing predictable randomness. Should the seed be set by the user as well?

Also, for the arbitrary matrix size, I'm considering adding a feature flag that is enabled when building the microbenchmarks directory, then, it will read the JSON file with the required args.

**Comment by ppenna** [comment]
> Just to clarify, does the matrix generation has to be done during the execution of the benchmark? If so, that would mess up the benchmark measurements with additional execution time that is non-related to the multiplication.

Yes, you should generate it during execution time. Do not worry about messing up with profiling statistics. Here is the rationale for that:

1. Matrix multiplication time will anyhow dominate the time for initializing matrices, as per its algorithmic complexity.
2. By dimensioning matrices at build time and filling them at runtime you will be able to study much more interesting things than having matrices with fixed size and static content.

> Also, for the arbitrary matrix size, I'm considering adding a feature flag that is enabled when building the microbenchmarks directory, then, it will read the JSON file with the required args.

You can do that in `build.rs` itself and not touch `Make` system at all.
